http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpFullRequest.java

```diff
@@ -66,9 +71,29 @@
         @Override
         default Builder uri(URI uri) {
             return this.protocol(uri.getScheme())
-                       .host(uri.getHost())
-                       .port(uri.getPort())
-                       .encodedPath(SdkHttpUtils.appendUri(uri.getRawPath(), encodedPath()));
+                    .host(uri.getHost())
+                    .port(uri.getPort())
+                    .encodedPath(SdkHttpUtils.appendUri(uri.getRawPath(), encodedPath()));
+        }
+
+        /**
+         * Convenience method to set the {@link #protocol()}, {@link #host()}, {@link #port()},
+         * {@link #encodedPath()} and keep query parameters if present from a {@link URI} object.
+         *
+         * @param uri URI containing protocol, host, port, path and optional query parameters.
+         * @return This builder for method chaining.
+         */
+        default Builder uriWithQueryParams(URI uri) {
+            Builder builder = this.uri(uri);
+            if (uri.getQuery() != null) {
+                Pattern.compile("\\s*&|%26\\s*")
+                       .splitAsStream(uri.getQuery().trim())
+                       .map(s -> s.split("=|%3D", 2))
+                       .collect(groupingBy(a -> a[0], mapping(a -> a[1], toList())))
+                       .forEach((paramKey, paramValues) -> paramValues
+                           .forEach(paramValue -> this.appendUriQueryParameter(paramKey, paramValue)));
```

Comment:
This code is pretty complex and is duplicated here and in `SdkHttpRequest`. Let's move this out into `SdkHttpUtils` instead. Should also make the parsing easier to test in isolation

Reply:
Agreed

Reply:
extracted